sources/api_client.py

```diff
@@ -0,0 +1,217 @@
+        https://docs.github.com/en/rest/guides/traversing-with-pagination
+    """
+
+    def __init__(self, links_next_key: str = "next") -> None:
+        """
+        Args:
+            links_next_key (str, optional): The key (rel ) in the 'Link' header
+                that contains the next page URL. Defaults to 'next'.
+        """
+        super().__init__()
+        self.links_next_key = links_next_key
+
+    def update_state(self, response: Response) -> None:
+        self.next_reference = response.links.get(self.links_next_key, {}).get("url")
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json()
+
+
+class JSONResponsePaginator(BaseNextUrlPaginator):
+    """A paginator that uses a specific key in the JSON response to find
+    the next page URL.
+    """
+
+    def __init__(self, next_key: str = "next", records_key: str = "results"):
+        """
+        Args:
+            next_key (str, optional): The key in the JSON response that
+                contains the next page URL. Defaults to 'next'.
+            records_key (str, optional): The key in the JSON response that
+                contains the page's records. Defaults to 'results'.
+        """
+        super().__init__()
+        self.next_key = next_key
+        self.records_key = records_key
+
+    def update_state(self, response: Response):
+        self.next_reference = response.json().get(self.next_key)
+
+    def extract_records(self, response: Response) -> Any:
+        return response.json().get(self.records_key, [])
+
+
+class BearerTokenAuth(AuthBase):
+    def __init__(self, token: str):
+        self.token = token
+
+    def __call__(self, request):
+        request.headers["Authorization"] = f"Bearer {self.token}"
+        return request
+
+
+def join_url(base_url: str, path: str) -> str:
+    if not base_url.endswith("/"):
+        base_url += "/"
+    return base_url + path.lstrip("/")
+
+
+class APIClient:
+    """A generic API client for making requests to an API.
+
+    Attributes:
+        base_url (str): The base URL of the API.
+        headers (Optional[Dict[str, str]]): Headers to include in all requests.
+        auth (Optional[AuthBase]): An authentication object to use for all requests.
+        paginator (Optional[BasePaginator]): A paginator object for handling API pagination.
+            Note that this object will be deepcopied for each request to ensure that the
+            paginator state is not shared between requests.
+    """
+
+    def __init__(
+        self,
+        base_url: str,
+        headers: Optional[Dict[str, str]] = None,
+        auth: Optional[AuthBase] = None,
+        paginator: Optional[BasePaginator] = None,
+    ) -> None:
+        self.base_url = base_url
+        self.headers = headers
+        self.auth = auth
+        self.paginator = paginator if paginator else HeaderLinkPaginator()
+
+    def make_request(self, path="", method="get", params=None, json=None):
```

Comment:
I think we need to start [requests.session](https://requests.readthedocs.io/en/latest/user/advanced/) here because they can be pre-configured with authentication strategy, custom headers etc.